kani-driver/src/cbmc_output_parser.rs

```diff
@@ -0,0 +1,991 @@
+                let line_str = format!(":{}", line);
+                fmt_str.push_str(line_str.as_str());
+                if let Some(column) = self.column.clone() {
+                    let column_str = format!(":{}", column);
+                    fmt_str.push_str(column_str.as_str());
+                }
+            }
+        } else {
+            fmt_str.push_str("Unknown file");
+        }
+        if let Some(function) = self.function.clone() {
+            let fun_str = format!(" in function {}", function);
+            fmt_str.push_str(fun_str.as_str());
+        }
+
+        write! {f, "{}", fmt_str}
+    }
+}
+
+/// Returns a path relative to the current working directory.
+fn filepath(file: String) -> String {
+    let file_path = PathBuf::from(file.clone());
+    let cur_dir = env::current_dir().unwrap();
+
+    let diff_path_opt = diff_paths(file_path, cur_dir);
+    if let Some(diff_path) = diff_path_opt {
+        diff_path.into_os_string().into_string().unwrap()
+    } else {
+        file
+    }
+}
+/// Struct that represents traces.
+///
+/// In general, traces may include more information than this, but this is not
+/// documented anywhere. So we ignore the rest for now.
+#[derive(Clone, Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct TraceItem {
+    pub thread: u32,
+    pub step_type: String,
+    pub hidden: bool,
+    pub source_location: Option<SourceLocation>,
+}
+
+#[derive(Copy, Clone, Debug, Deserialize, PartialEq)]
+#[serde(rename_all = "UPPERCASE")]
+pub enum CheckStatus {
+    Failure,
+    Success,
+    Undetermined,
+    Unreachable,
+}
+
+impl std::fmt::Display for CheckStatus {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let check_str = match self {
+            CheckStatus::Success => style("SUCCESS").green(),
+            CheckStatus::Failure => style("FAILURE").red(),
+            CheckStatus::Unreachable => style("UNREACHABLE").yellow(),
+            CheckStatus::Undetermined => style("UNDETERMINED").yellow(),
+        };
+        write! {f, "{}", check_str}
+    }
+}
+
+#[derive(PartialEq)]
+enum Action {
+    ClearInput,
+    ProcessItem,
+}
+
+/// A parser for CBMC output, whose state is determined by:
+///  1. The input accumulator, required to process items on the fly.
+///  2. The buffer, which is accessed to retrieve more lines.
+///
+/// CBMC's JSON output is defined as a JSON array which contains:
+///  1. One program at the beginning (i.e., a message with CBMC's version).
+///  2. Messages, which can appears anywhere, and are either status messages or error messages.
+///  3. Verification results, another JSON array with all individual checks.
+///  4. Prover status, at the end. Because the verification results depends on
+///     our postprocessing, this is not used.
+///
+/// The parser reads the output line by line. A line may trigger one action, and
+/// the action may return a parsed item.
+struct Parser<'a, 'b> {
```

Comment:
Would using something like the [StreamDeserializer](https://docs.rs/serde_json/latest/serde_json/struct.StreamDeserializer.html#) be possible here?

Reply:
Good point.

I tried using `StreamDeserializer` for parsing from the buffer in one of my earliest attempts, but it wasn't successful. The problem is that it expects self-delineating values, but the output we process is a JSON array with delineated values. In other words, the output we get is of the form:
```
[
    Program,
    Message,
    ...,
    Message,
    Result,
    Message,
    ProverStatus
]
```
But for `StreamDeserializer` to work we would need an output of the form:
```
Program
Message
...
Message
Result
Message
ProverStatus
```

Note that the parsing logic is mostly concerned with this problem: It assumes it's already within a JSON array, ignoring its delimiters. Changing the CBMC output to have the second form described is a no-go, at least according to the discussions I had then.

Reply:
I see.